liblsp/InputHandler.cpp

```diff
@@ -0,0 +1,223 @@
+	else
+		return nullopt;
+}
+
+optional<ShutdownParams> InputHandler::shutdown(MessageId const&, Json::Value const&)
+{
+	m_shutdownRequested = true;
+	return ShutdownParams{};
+}
+
+optional<ExitParams> InputHandler::exit(MessageId const&, Json::Value const&)
+{
+	return ExitParams{};
+}
+
+optional<InitializeRequest> InputHandler::initializeRequest(MessageId const& _id, Json::Value const& _args)
+{
+	lsp::protocol::InitializeRequest request{};
+	request.requestId = _id;
+
+	if (Json::Value pid = _args["processId"]; pid)
+		request.processId = pid.asInt();
+
+	if (Json::Value rootPath = _args["rootPath"]; rootPath)
+		request.rootPath = rootPath.asString();
+
+	if (Json::Value uri = _args["rootUri"]; uri)
+		request.rootUri = uri.asString();
+
+	if (Json::Value trace = _args["trace"]; trace)
+	{
+		string const name = trace.asString();
+		if (name == "messages")
+			request.trace = lsp::protocol::Trace::Messages;
+		else if (name == "verbose")
+			request.trace = lsp::protocol::Trace::Verbose;
+		else if (name == "off")
+			request.trace = lsp::protocol::Trace::Off;
+	}
+
+	if (Json::Value folders = _args["workspaceFolders"]; folders)
+	{
+		for (Json::Value folder: folders)
+		{
+			lsp::protocol::WorkspaceFolder wsFolder{};
+			wsFolder.name = folder["name"].asString();
+			wsFolder.uri = folder["uri"].asString();
+			request.workspaceFolders.emplace_back(move(wsFolder));
+		}
+	}
+
+	// TODO: initializationOptions
+	// TODO: ClientCapabilities
+
+	return request;
+}
+
+optional<protocol::InitializedNotification> InputHandler::initialized(MessageId const&, Json::Value const&)
+{
+	// TODO: error checking?
+	return InitializedNotification{};
+}
+
+optional<DidOpenTextDocumentParams> InputHandler::textDocument_didOpen(MessageId const& _id, Json::Value const& _args)
+{
+	if (!_args["textDocument"])
+		return nullopt;
+
+	DidOpenTextDocumentParams args{};
+	args.requestId = _id;
+	args.textDocument.uri = _args["textDocument"]["uri"].asString();
+	args.textDocument.languageId = _args["textDocument"]["languageId"].asString();
+	args.textDocument.version = _args["textDocument"]["version"].asInt();
+	args.textDocument.text = _args["textDocument"]["text"].asString();
+
+	return args;
+}
+
+optional<protocol::DidChangeTextDocumentParams> InputHandler::textDocument_didChange(MessageId const& _id, Json::Value const& _json)
+{
+	DidChangeTextDocumentParams didChange{};
+	didChange.requestId = _id;
+	didChange.textDocument.version = _json["textDocument"]["version"].asInt();
+	didChange.textDocument.uri = _json["textDocument"]["uri"].asString();
+
+	for (Json::Value jsonContentChange: _json["contentChanges"])
+	{
+		if (jsonContentChange.isObject() && jsonContentChange["range"])
+		{
+			TextDocumentRangedContentChangeEvent rangedChange;
+			rangedChange.text = jsonContentChange["text"].asString();
+
+			Json::Value jsonRange = jsonContentChange["range"];
+
+			rangedChange.range.start.line = jsonRange["start"]["line"].asInt();
```

Comment:
Do you have any kind of error handling? `asInt` and the like will assert on invalid data.

Reply:
Perhaps you need to catch JSON errors and return some "mismatched input" message to the editor. See the StandardCompiler.

Reply:
I think there is a global try-catch in the outermost possible I/O loop to report internal errors during request handling. This should catch it.